### PR TITLE
test

### PR DESCRIPTION
**Pull request by charles-cowart**

### CODE REVIEW

**Comment by charles-cowart** [comment]
@antgonza Ready for review!

File: qp_klp/process_amplicon_job.py [hunk 1]
```diff
@@ -262,24 +263,64 @@ def process_amplicon(mapping_file_path, qclient, run_identifier, out_dir,
...
 
-    map_sample_names_to_tube_ids(pfp_list, sn_tid_map_by_project)
+        # if seq_pro is run, prep_file_paths will be populated by run().
+        for study_id in gpf_job.prep_file_paths:
+            for prep_file_path in gpf_job.prep_file_paths[study_id]:
+                metadata_dict = pd.read_csv(prep_file_path,
+                                            delimiter='\t').to_dict('index')
+
+                # determine data_type based on target_gene column.
+                for key in {'16S': '16S', '18S': '18S', 'ITS': 'ITS'}:
```
Comment (antgonza) [inline qp_klp/process_amplicon_job.py:283]: This should be a set, not a dict ... as AFAIK the comparison will only work on the keys.

File: qp_klp/process_metagenomics_job.py [hunk 2]
```diff
@@ -244,24 +247,58 @@ def process_metagenomics(sample_sheet_path, lane_number, qclient,
...
+                data = {'prep_info': dumps(metadata_dict),
+                        'study': study_id,
+                        'data_type': assay_type}
+
+                reply = qclient.post('/apitest/prep_template/', data=data)
+                prep_id = reply['prep']
+
+                if study_id not in touched_studies_prep_info:
+                    touched_studies_prep_info[study_id] = []
+                touched_studies_prep_info[study_id].append(prep_id)
```
Comment (antgonza) [inline qp_klp/process_metagenomics_job.py:276]: DefaultDict

File: qp_klp/process_amplicon_job.py [hunk 3]
```diff
@@ -262,24 +263,64 @@ def process_amplicon(mapping_file_path, qclient, run_identifier, out_dir,
...
+                data = {'prep_info': dumps(metadata_dict),
+                        'study': study_id,
+                        'data_type': data_type}
+
+                reply = qclient.post('/apitest/prep_template/', data=data)
+                prep_id = reply['prep']
+
+                if study_id not in touched_studies_prep_info:
+                    touched_studies_prep_info[study_id] = []
+                touched_studies_prep_info[study_id].append(prep_id)
```
Comment (antgonza) [inline qp_klp/process_amplicon_job.py:297]: DefaultDict

**Review by antgonza — CHANGES_REQUESTED** [review]
Looking great, thank you.

**Comment by charles-cowart** [comment]
@antgonza ready for review!

**Review by antgonza — APPROVED** [review]
Looks good, thank you. I'll deploy and test in a few hours.

File: qp_klp/process_amplicon_job.py [hunk 4]
```diff
@@ -262,24 +264,62 @@ def process_amplicon(mapping_file_path, qclient, run_identifier, out_dir,
...
+                for key in {'16S', '18S', 'ITS'}:
+                    if key in metadata_dict[0]['target_gene']:
+                        data_type = key
+                        break
+
+                data = {'prep_info': dumps(metadata_dict),
+                        'study': study_id,
+                        'data_type': data_type}
+
+                reply = qclient.post('/apitest/prep_template/', data=data)
```
Comment (antgonza) [inline qp_klp/process_amplicon_job.py:293]: This should be `reply = qclient.post('/qiita_db/prep_template/', data=data)`

File: qp_klp/process_amplicon_job.py [hunk 5]
```diff
@@ -262,24 +264,62 @@ def process_amplicon(mapping_file_path, qclient, run_identifier, out_dir,
...
+        # if seq_pro is run, prep_file_paths will be populated by run().
+        for study_id in gpf_job.prep_file_paths:
+            for prep_file_path in gpf_job.prep_file_paths[study_id]:
+                metadata_dict = pd.read_csv(prep_file_path,
+                                            delimiter='\t').to_dict('index')
 
-    map_sample_names_to_tube_ids(pfp_list, sn_tid_map_by_project)
+                # determine data_type based on target_gene column.
+                for key in {'16S', '18S', 'ITS'}:
+                    if key in metadata_dict[0]['target_gene']:
```
Comment (antgonza) [inline qp_klp/process_amplicon_job.py:285]: This should be: `if key in metadata_dict[list(metadata_dict.keys())[0]]['target_gene']:`

**Review by antgonza — CHANGES_REQUESTED** [review]
A few comments on the aplicon code that should also be reflected in the metagenomic (what I mean is that I only made comments on the amplicon but those changes should also be applied in the metagenomic. Thank you.

File: qp_klp/process_amplicon_job.py [hunk 6]
```diff
@@ -262,24 +264,62 @@ def process_amplicon(mapping_file_path, qclient, run_identifier, out_dir,
...
-                if prep_file.endswith('.tsv'):
-                    pfp_list.append(join(root, prep_file))
+        # list. Replace sample-names w/tube-ids in all relevant prep-files.
+        preps = list(chain.from_iterable(gpf_job.prep_file_paths.values()))
+        map_sample_names_to_tube_ids(preps, sn_tid_map_by_project)
+
+        # if seq_pro is run, prep_file_paths will be populated by run().
+        for study_id in gpf_job.prep_file_paths:
+            for prep_file_path in gpf_job.prep_file_paths[study_id]:
+                metadata_dict = pd.read_csv(prep_file_path,
```
Comment (antgonza) [inline qp_klp/process_amplicon_job.py:280]: Can you make sure the variable name doesn't have the variable type in the name?

File: qp_klp/process_amplicon_job.py (hunk 5 above)
Comment (antgonza) [reply qp_klp/process_amplicon_job.py:285]: I don't follow your question but just to clear the issue is `metadata_dict` is a dict and not a dataframe so we need to get the first key of the dict, then get the 'target_gene' of that key.

Comment (charles-cowart) [reply qp_klp/process_amplicon_job.py]: Ah I see what you're saying, ty!

File: qp_klp/process_metagenomics_job.py [hunk 7]
```diff
@@ -244,24 +248,57 @@ def process_metagenomics(sample_sheet_path, lane_number, qclient,
...
+            'SKD8.640184': {'primer': 'GTGCCAGCMGCCGCGGTAA',
+                            'barcode': 'GTCCGCAAGTTA',
+                            'platform': 'Illumina',
+                            'instrument_model': 'Illumina MiSeq'}}
+
+        data = {'prep_info': dumps(metadata),
+                'study': '1',
+                'data_type': 'Metagenomic'}
+
+        reply = qclient.post('/qiita_db/prep_template/', data=data)
```
Comment (antgonza) [inline qp_klp/process_metagenomics_job.py:299]: This still needs to be changed!

**Comment by charles-cowart** [comment]
Adding in fix originally addressed by https://github.com/qiita-spots/qp-knight-lab-processing/pull/46 using the better alternative to random.choice() (random.sample()).

**Comment by charles-cowart** [comment]
@antgonza ready for review!

File: qp_klp/process_amplicon_job.py [hunk 8]
```diff
@@ -262,24 +265,65 @@ def process_amplicon(mapping_file_path, qclient, run_identifier, out_dir,
...
+            'SKD8.640184': {'primer': 'GTGCCAGCMGCCGCGGTAA',
+                            'barcode': 'GTCCGCAAGTTA',
+                            'platform': 'Illumina',
+                            'instrument_model': 'Illumina MiSeq'}}
+
+        data = {'prep_info': dumps(metadata),
+                'study': '1',
+                'data_type': '16S'}
+
+        reply = qclient.post('/apitest/prep_template/', data=data)
```
Comment (antgonza) [inline qp_klp/process_amplicon_job.py:324]: This should be `reply = qclient.post('/qiita_db/prep_template/', data=data)`, no?

**Review by antgonza — APPROVED** [review]

**Pull request merged**